network/v1alpha1/types_dnsnameresolver.go

```diff
@@ -0,0 +1,108 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +openshift:compatibility-gen:level=4
+
+// DNSNameResolver stores the DNS name resolution information of a DNS name. It is TechPreviewNoUpgrade only.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+type DNSNameResolver struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the DNSNameResolver.
+	// +kubebuilder:validation:Required
+	Spec DNSNameResolverSpec `json:"spec"`
+	// status is the most recently observed status of the DNSNameResolver.
+	// +optional
+	Status DNSNameResolverStatus `json:"status,omitempty"`
+}
+
+// DNSNameResolverSpec is a desired state description of DNSNameResolver.
+type DNSNameResolverSpec struct {
+	// name is the DNS name for which the DNS name resolution information will be stored.
+	// For a regular DNS name, only the DNS name resolution information of the regular DNS
+	// name will be stored. For a wildcard DNS name, the DNS name resolution information
+	// of all the DNS names, that matches the wildcard DNS name, will be stored.
+	// For a wildcard DNS name, the '*' will match only one label. Additionally, only a single
+	// '*' can be used at the beginning of the wildcard DNS name. For example, '*.example.com.'
+	// will match 'sub1.example.com.' but won't match 'sub2.sub1.example.com.'
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="spec.name is immutable"
+	Name string `json:"name"`
```

Comment:
I will add the maximum length limit as 254 because we have the trailing period: https://devblogs.microsoft.com/oldnewthing/?p=7873

Reply:
Not sure it will matter, but isn't the guidance to use the constraints in kubernetes/apimachinery (253 in this case):  https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L209

Reply:
It appears the RegEx now only accepts patterns ending in a trailing dot.

Reply:
> Not sure it will matter, but isn't the guidance to use the constraints in kubernetes/apimachinery (253 in this case): https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L209

@JoelSpeed should have better understanding of what is the correct approach on this.

> It appears the RegEx now only accepts patterns ending in a trailing dot, is that intentional?

Yes. If the trailing dot is not made mandatory, then there is a chance that the same DNS name may have 2 DNSNameResolver CRs, one with a trailing period and another without it. To avoid this, the trailing period is made mandatory.

Reply:
I think the 254 limit is acceptable since I was expected the validation from API machinery as Candace has pointed out, but adding the trailing dot adds an extra character.

Otherwise I thought we were compliant with the DNS1123 spec?

Reply:
> Otherwise I thought we were compliant with the DNS1123 spec?

DNS1123 spec also requires the DNS names to be in lowercase. Currently, both lowercase and uppercase characters are valid. I can remove the support for uppercase character as anyway these CRs will be created by OVN-K cluster manager after changing everything to lowercase and adding the trailing period if it is missing.

Reply:
I have removed the support of uppercase letters and added the validation for 63 characters for each label. I have added integration tests for the same as well.

Reply:
In the end, having the validation here is extra, because the component that creates this should do it first.  A user may be trying to lookup www.Example.com, but we wouldn't report that www.Example.com is invalid, or doesn't exist.  We would just lowercase it and lookup www.example.com for them, I think?

Reply:
Yes. That's true. But it is good that the validation exists. The components which are going to read the field and use it wouldn't have to do the extra validation as it is already validated. Additionally, in the future the DNSNameResolver can be used for other purposes as well apart from EgressFirewall. Having this validation in place will help avoid possible issues in future.

Reply:
Validation should be in place even if it duplicates existing validation. While the API exists in the cluster, it's exposed to end users. And while they shouldn't be modifying it, they might. So it's best to prevent them from committing garbage to etcd even if they shouldn't be touching it at all. This should minimise the blast radius of anything they try to do